### PR TITLE
[avmfritz] Switched to auto-generated DS descriptor and more

### PR DESCRIPTION
**Pull request by cweitkamp**
- Switched to auto-generated DS descriptor
- Added support for configurable background discovery
- Added vendor property of discovered devices
- Fixed stopping background discovery service and other minor optimizations

Signed-off-by: Christoph Weitkamp <github@christophweitkamp.de>

### CODE REVIEW

File: addons/binding/org.openhab.binding.avmfritz/src/main/java/org/openhab/binding/avmfritz/internal/discovery/AVMFritzUpnpDiscoveryParticipant.java [hunk 1]
```diff
@@ -23,22 +23,24 @@
...
- * The {@link AvmUpnpDiscoveryParticipant} is responsible for discovering
+ * The {@link AVMFritzUpnpDiscoveryParticipant} is responsible for discovering
  * new and removed FRITZ!Box devices. It uses the central
  * {@link UpnpDiscoveryService}.
  *
  * @author Robert Bausdorf - Initial contribution
  * 
  */
-public class AvmUpnpDiscoveryParticipant implements UpnpDiscoveryParticipant {
+@Component(service = UpnpDiscoveryParticipant.class, immediate = true, configurationPid = "discovery.avmfritz")
```
Comment (kaikreuzer) [inline addons/binding/org.openhab.binding.avmfritz/src/main/java/org/openhab/binding/avmfritz/internal/discovery/AVMFritzUpnpDiscoveryParticipant.java]: This is a UPnP discovery participant - it does not make sense to pass configurations with pid "discovery.avmfritz" to it as it itself does not implement any background discovery (it is the UPnP service, which does it).

File: addons/binding/org.openhab.binding.avmfritz/src/main/java/org/openhab/binding/avmfritz/internal/discovery/AVMFritzDiscoveryService.java [hunk 2]
```diff
@@ -79,12 +80,27 @@ public AVMFritzDiscoveryService(BoxHandler bridgeHandler) {
         this.activate(null);
     }
 
+    @Override
+    protected void activate(Map<String, Object> configProperties) {
```
Comment (kaikreuzer) [inline addons/binding/org.openhab.binding.avmfritz/src/main/java/org/openhab/binding/avmfritz/internal/discovery/AVMFritzDiscoveryService.java]: I'd assume that this method is never called as this class is not handled by DS.

File: addons/binding/org.openhab.binding.avmfritz/src/main/java/org/openhab/binding/avmfritz/internal/discovery/AVMFritzDiscoveryService.java [hunk 3]
```diff
@@ -79,12 +80,27 @@ public AVMFritzDiscoveryService(BoxHandler bridgeHandler) {
...
+        super.activate(configProperties);
+    }
+
+    @Override
+    protected void modified(Map<String, Object> configProperties) {
```
Comment (kaikreuzer) [inline addons/binding/org.openhab.binding.avmfritz/src/main/java/org/openhab/binding/avmfritz/internal/discovery/AVMFritzDiscoveryService.java]: I'd assume that this method is never called as this class is not handled by DS.

File: addons/binding/org.openhab.binding.avmfritz/src/main/java/org/openhab/binding/avmfritz/internal/discovery/AVMFritzDiscoveryService.java [hunk 4]
```diff
@@ -79,12 +80,27 @@ public AVMFritzDiscoveryService(BoxHandler bridgeHandler) {
...
+        super.modified(configProperties);
+    }
+
+    public void deactivate() {
+        super.deactivate();
+        removeOlderResults(new Date().getTime());
```
Comment (kaikreuzer) [inline addons/binding/org.openhab.binding.avmfritz/src/main/java/org/openhab/binding/avmfritz/internal/discovery/AVMFritzDiscoveryService.java]: Why should valid results be removed here?

File: addons/binding/org.openhab.binding.avmfritz/ESH-INF/config/config.xml [hunk 5]
```diff
@@ -42,7 +42,7 @@
 			<description>User name which has HomeAuto permissions on the given FRITZ!Box.
 			</description>
 		</parameter>
-		<parameter name="password" type="text" required="false" groupName="authentication">
+		<parameter name="password" type="text" required="true" groupName="authentication">
```
Comment (kaikreuzer) [inline addons/binding/org.openhab.binding.avmfritz/ESH-INF/config/config.xml]: Afair, the password on a FritzBox has been optional so far. Has this changed?

File: addons/binding/org.openhab.binding.avmfritz/ESH-INF/config/config.xml [hunk 6]
```diff
@@ -104,7 +104,7 @@
 				<option value="https">HTTPS</option>
 			</options>
 		</parameter>
-		<parameter name="password" type="text" required="false" groupName="authentication">
+		<parameter name="password" type="text" required="true" groupName="authentication">
```
Comment (kaikreuzer) [inline addons/binding/org.openhab.binding.avmfritz/ESH-INF/config/config.xml]: Afair, the password on a FritzBox has been optional so far. Has this changed?

File: addons/binding/org.openhab.binding.avmfritz/ESH-INF/config/config.xml (hunk 5 above)
Comment (cweitkamp) [reply addons/binding/org.openhab.binding.avmfritz/ESH-INF/config/config.xml]: No, not really. But there are different places in the code (see [BoxHandler.java#L104](https://github.com/cweitkamp/openhab2-addons/blob/eb7518303774845dc669d1a974ccb7b0a78a6deb/addons/binding/org.openhab.binding.avmfritz/src/main/java/org/openhab/binding/avmfritz/handler/BoxHandler.java#L104) and [FritzahaWebInterface.java#L92](https://github.com/cweitkamp/openhab2-addons/blob/eb7518303774845dc669d1a974ccb7b0a78a6deb/addons/binding/org.openhab.binding.avmfritz/src/main/java/org/openhab/binding/avmfritz/internal/hardware/FritzahaWebInterface.java#L92)) where the binding stops further progress when no password is set. So I suggest we better set it to required.

File: addons/binding/org.openhab.binding.avmfritz/src/main/java/org/openhab/binding/avmfritz/internal/discovery/AVMFritzDiscoveryService.java (hunk 2 above)
Comment (cweitkamp) [reply addons/binding/org.openhab.binding.avmfritz/src/main/java/org/openhab/binding/avmfritz/internal/discovery/AVMFritzDiscoveryService.java]: Maybe I misunderstood something, but didn't we register this discovery as a new service once a BoxHandler is created (see [HandlerFactory.java#L69](https://github.com/cweitkamp/openhab2-addons/blob/eb7518303774845dc669d1a974ccb7b0a78a6deb/addons/binding/org.openhab.binding.avmfritz/src/main/java/org/openhab/binding/avmfritz/internal/HandlerFactory.java#L69))?

Comment (kaikreuzer) [reply addons/binding/org.openhab.binding.avmfritz/src/main/java/org/openhab/binding/avmfritz/internal/discovery/AVMFritzDiscoveryService.java]: Yes, but you must not confuse the registration of a class instance as an OSGi service with declaring a service and handing over the lifecycle to DS.

File: addons/binding/org.openhab.binding.avmfritz/src/main/java/org/openhab/binding/avmfritz/internal/discovery/AVMFritzDiscoveryService.java (hunk 4 above)
Comment (cweitkamp) [reply addons/binding/org.openhab.binding.avmfritz/src/main/java/org/openhab/binding/avmfritz/internal/discovery/AVMFritzDiscoveryService.java]: In which situations will this service be deactivated? I can think of two: First the user deactivates it in his environment for a reason and second the binding gets unistalled. Are there more? And why should we leave them there?

Comment (kaikreuzer) [reply addons/binding/org.openhab.binding.avmfritz/src/main/java/org/openhab/binding/avmfritz/internal/discovery/AVMFritzDiscoveryService.java]: Well, if the service is shut down, if some parts are updated and bundle restart, etc.
The Inbox persists its entries, as long as they are valid, they simply should be kept as they are.

Comment (cweitkamp) [reply addons/binding/org.openhab.binding.avmfritz/src/main/java/org/openhab/binding/avmfritz/internal/discovery/AVMFritzDiscoveryService.java]: You convinced me. Should I remove it in the `stopScan()` method too? ([AVMFritzDiscoveryService.java#L105](https://github.com/cweitkamp/openhab2-addons/blob/ad8b3ea1808903f1ac890335c6507cc2da081937/addons/binding/org.openhab.binding.avmfritz/src/main/java/org/openhab/binding/avmfritz/internal/discovery/AVMFritzDiscoveryService.java#L105))

Comment (kaikreuzer) [reply addons/binding/org.openhab.binding.avmfritz/src/main/java/org/openhab/binding/avmfritz/internal/discovery/AVMFritzDiscoveryService.java]: yes, I would think so.

Comment (cweitkamp) [reply addons/binding/org.openhab.binding.avmfritz/src/main/java/org/openhab/binding/avmfritz/internal/discovery/AVMFritzDiscoveryService.java]: done

**Comment by kaikreuzer** [comment]
Thanks!

**Pull request merged**